### PR TITLE
[reconfigurator] introduce a simulator, make CLI use it

### PR DESCRIPTION
**Pull request by sunshowers**
For randomized testing, it's useful to have a higher-level notion of a
succession of system states, which can be inspected, rewound, and other
branches explored as desired. Introduce this via a new `Simulator` struct.

The individual states are roughly identical to the existing
`ReconfiguratorSim` in the CLI, and are a pretty standard tree structure. For
now I've chosen not to use a Merkle tree, just a UUID-based tree, but if
there's a good use case for it we may choose to make it a Merkle tree in the
future.

I've added a few basic features like storing the log of changes (though it's
not complete... I think I'll want to make `SystemDescription` follow the same
structure with read-only and mutable versions.)

As a proof of concept, this PR also rebuilds the current reconfigurator-cli
script on top of the simulator. I've not made any functional changes to the
CLI, but hope to add features like looking at the tree of states and undoing
them, and with non-deterministic testing to be able to load up the tree of
states.

The simulator uses a bunch of structural sharing via `Arc`. I was wondering
whether I could use the `im` crate, but it doesn't seem to have an `IndexMap`
equivalent. The `Arc`s should make cloning cheap enough that I don't think it's
an issue in any case. (There's a comment in the PR about why I haven't used an
object-store-like style instead -- it's just easier to use `Arc` for now.)

There's a bunch of work to be done before this can be landed, but I wanted to
put it up for folks to have an early look.

I've left a few TODOs in for future work. I don't think any of them block this
PR.

### CODE REVIEW

File: nexus/reconfigurator/simulation/src/sim.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    /// Get the current heads of the store.
+    #[inline]
+    pub fn heads(&self) -> &IndexSet<ReconfiguratorSimUuid> {
+        &self.heads
+    }
+
+    /// Get the state for the given UUID.
+    pub fn get_state(&self, id: ReconfiguratorSimUuid) -> Option<&SimState> {
+        if id == Self::ROOT_ID {
+            return Some(&self.root_state);
```
Comment (jgallagher) [inline nexus/reconfigurator/simulation/src/sim.rs:103]: This seems fine, but out of curiosity - why not insert the root state into `self.states`?

File: nexus/reconfigurator/simulation/src/state.rs [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+    root_state: SyncConstPtr<SimState>,
+    id: ReconfiguratorSimUuid,
+    // The parent state that this state was derived from.
+    parent: Option<ReconfiguratorSimUuid>,
+    // The state's generation, starting from 0.
+    //
+    // TODO: Should this be its own type to avoid confusion with other
+    // Generation instances? Generation numbers start from 1, but in our case 0
+    // provides a better user experience.
+    generation: Generation,
```
Comment (andrewjstone) [inline nexus/reconfigurator/simulation/src/state.rs:42]: I'd probably change this to `Version` to avoid confusion.

File: nexus/reconfigurator/simulation/src/system.rs [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+    }
+
+    pub fn get_internal_dns(
+        &self,
+        generation: Generation,
+    ) -> Result<&DnsConfigParams, KeyError> {
+        self.internal_dns
+            .get(&generation)
+            .map(|d| &**d)
+            .ok_or_else(|| KeyError::internal_dns(generation))
```
Comment (jgallagher) [inline nexus/reconfigurator/simulation/src/system.rs:143]: I don't really have a preference, but `get_{collection,blueprint}` use a `match` and `get_{internal,external}_dns` use `.map().ok_or_else()` to do the same thing, I think? Probably would be less surprising if they all used the same style.

File: nexus/reconfigurator/simulation/src/system.rs [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+    //    UUID. Unlike (say) source control commit hashes, UUIDs are not
+    //    content-hashed, so the same UUID can be associated with different
+    //    blueprints/collections.
+    // 2. DNS configs are absolutely not unique per generation! Again, not
+    //    content-hashed.
+    // 3. The mutable system may wish to not add objects to the store until
+    //    it's committed. This means that the mutable system would probably
+    //    have to maintain a list of pending objects to add to the store. That
+    //    complicates some of the internals, if not the API.
+    // 4. We'll have to figure out how to manage the store (so SimSystemBuilder
```
Comment (andrewjstone) [inline nexus/reconfigurator/simulation/src/system.rs:47]: It's unclear to me why any of this code would need to be multithreaded, especially if it's fully deterministic. The only use I could really see for multi-threading is computationally expensive things where the computation is deterministic. For things like that we could call out to something like `rayon` if necessary.

File: nexus/reconfigurator/simulation/src/system.rs [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+    }
+
+    fn add_internal_dns_inner(
+        &mut self,
+        params: Arc<DnsConfigParams>,
+    ) -> Result<(), DuplicateError> {
+        let generation = params.generation;
+        match self.system.internal_dns.entry(generation) {
+            std::collections::btree_map::Entry::Vacant(entry) => {
+                entry.insert(params);
```
Comment (jgallagher) [inline nexus/reconfigurator/simulation/src/system.rs:391]: Should the internal/external DNS operations get entries in `self.log`?

File: nexus/reconfigurator/simulation/src/state.rs [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+                            baseboard_id: &baseboard_id,
+                            sp: inv_sp,
+                            rot: inv_rot,
+                        })
+                    } else {
+                        None
+                    }
+                });
+
+            // XXX: Should this error ever happen? The only case where it
```
Comment (jgallagher) [inline nexus/reconfigurator/simulation/src/state.rs:292]: I'd maybe move this comment down into the `match`? On first read I wasn't sure what error it was talking about, since there's no `?` and the method is documented as infallible.

**Review by jgallagher — APPROVED** [review]
Looks good to me! Just a few nits / questions.

File: nexus/reconfigurator/simulation/src/state.rs [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+            id,
+            description,
+            parent: Some(self.parent),
+            generation: self.parent_gen.next(),
+            system,
+            config,
+            rng,
+            log,
+        };
+        sim.add_state(Arc::new(state));
```
Comment (andrewjstone) [inline nexus/reconfigurator/simulation/src/state.rs:425]: There are a bunch of comments around why things look this way, but I'm still not sold that it's the right way to go. For example, there's a requirement that `sim.next_sim_uuid()` is called before `sim.add_state()`, but this cannot be checked at compile time. I think rather than passing in a `Simulator` to this method, the `Simulator` should have a method to return a mutable reference to the "next" `SimStateBuilder`, or even better, take a closure that can generate the next `SimState` when passed an `&SimStateBuilder`. Then when that closure returns the  committed state can be added by the simulator.

In short, I'd like this all to be "driven" via the `Simulator` without having to pass in the simulator. 

It's totally possible I'm missing something here, but I think there may be room to simplify the API a bit and would like to discuss more.

**Review by andrewjstone — COMMENTED** [review]
This is awesome @sunshowers! I Would love to walk through it a bit live and chat about goals and design.

**Review by andrewjstone — APPROVED** [review]
Chatted offline with @sunshowers. All my comments are minor. More than happy to see this merged and build upon it!

File: dev-tools/reconfigurator-cli/src/main.rs [hunk 8]
```diff
@@ -661,21 +618,26 @@ fn cmd_sled_add(
...
     add: SledAddArgs,
 ) -> anyhow::Result<Option<String>> {
-    let mut new_sled = SledBuilder::new();
-    if let Some(sled_id) = add.sled_id {
-        new_sled = new_sled.id(sled_id);
-    }
+    let mut state = sim.current_state().to_mut();
+    let sled_id = add.sled_id.unwrap_or_else(|| state.rng_mut().next_sled_id());
+    let new_sled = SledBuilder::new().id(sled_id);
+    state.system_mut().description_mut().sled(new_sled)?;
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dev-tools/reconfigurator-cli/src/main.rs:624]: This is the old name which I haven't changed in this PR -- I agree with you, we need to make a pass to make all the names indicate setters.

File: nexus/reconfigurator/simulation/src/sim.rs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    /// This is always defined to be the nil UUID, and if queried will always
+    /// have a state associated with it.
+    pub const ROOT_ID: ReconfiguratorSimUuid = ReconfiguratorSimUuid::nil();
+
+    /// Create a new simulator with the given initial seed.
+    pub fn new(log: &slog::Logger, seed: Option<String>) -> Self {
+        let seed = match seed {
+            Some(seed) => seed,
+            None => seed_from_entropy(),
+        };
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply nexus/reconfigurator/simulation/src/sim.rs:68]: Thanks, fixed.

File: nexus/reconfigurator/simulation/src/sim.rs (hunk 1 above)
Comment (sunshowers) [reply nexus/reconfigurator/simulation/src/sim.rs:103]: Seemed cleaner to me that way -- it could live in `self.states` instead of being its own member.

File: nexus/reconfigurator/simulation/src/system.rs (hunk 3 above)
Comment (sunshowers) [reply nexus/reconfigurator/simulation/src/system.rs:143]: Agreed, changed everything to `match`.

File: nexus/reconfigurator/simulation/src/system.rs (hunk 5 above)
Comment (sunshowers) [reply nexus/reconfigurator/simulation/src/system.rs:391]: Ah so there's a distinction between internal and external callers here -- external callers add to logs while internal callers don't. But I realized the code wasn't clear about this at all. I've refactored it into having two layers: the external public API layer which does logging, and the internal layer which does not.

File: nexus/reconfigurator/simulation/src/state.rs (hunk 6 above)
Comment (sunshowers) [reply nexus/reconfigurator/simulation/src/state.rs:292]: Good idea! I've done that and combined it with the existing comment below.

File: nexus/reconfigurator/simulation/src/sim.rs [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+/// hashes.) This does add some complexity, though, and there isn't a clear
+/// benefit at the moment. For more, see the comment within the definition of
+/// `system.rs`'s `SimSystem`.
+#[derive(Clone, Debug)]
+pub struct Simulator {
+    log: slog::Logger,
+    // In the future, it would be interesting to store a chain of every set of
+    // heads over time, similar to `jj op log`. That would let us implement undo
+    // and restore operations.
+    heads: IndexSet<ReconfiguratorSimUuid>,
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply nexus/reconfigurator/simulation/src/sim.rs:42]: Added -- it's the set of terminal nodes, similar to the list of git branches or jj/hg heads.

File: nexus/reconfigurator/simulation/src/config.rs [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    silo_names: IndexSet<Name>,
+
+    /// External DNS zone name configured
+    external_dns_zone_name: String,
+
+    /// The number of Nexus zones to create.
+    ///
+    /// TODO: This doesn't quite fit in here because it's more of a policy
+    /// setting than a config option. But we can't set it in the
+    /// `SystemDescription` because need to persist policy across system wipes.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply nexus/reconfigurator/simulation/src/config.rs:27]: Fuixed, thanks.

File: nexus/reconfigurator/simulation/src/state.rs (hunk 7 above)
Comment (sunshowers) [reply nexus/reconfigurator/simulation/src/state.rs:425]: As we discussed, I would prefer to keep it this way because it means that if the caller decides to throw away an intermediate state, they can easily do so. This is the general model I've seen other tools follow, and while it's slightly looser at compile time I think the flexibility is worth it.

There are some dependencies that can't easily be checked at compile-time, but only one that's externally visible: the simulator passed in must be the one the state originated from. I don't think this is a major burden.

> For example, there's a requirement that sim.next_sim_uuid() is called before sim.add_state(), but this cannot be checked at compile time.

This is true, but it's not externally visible -- this is managed purely within the reconfigurator-sim code.

**Comment by sunshowers** [comment]
Accidentally broke main -- see #7068 and #7069.
